SONARPHP-710 Rule: The output of functions that don't return anything should not be used

alban-auzeill opened the pull request.

SonarTech commented:
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

SonarTech commented on php-checks/src/main/java/org/sonar/php/checks/UseOfEmptyReturnValueCheck.java:
```diff
@@ -0,0 +1,576 @@
+  }
+
+  @Nullable
+  private String nameOf(Tree tree) {
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make "nameOf" a "static" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

pynicolas commented on php-checks/src/main/java/org/sonar/php/checks/UseOfEmptyReturnValueCheck.java:
```diff
@@ -0,0 +1,375 @@
+  }
+
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
```
What about `case function_call()`? We should rather create PHPTree.getParent().

SonarTech commented on php-checks/src/main/java/org/sonar/php/checks/UseOfEmptyReturnValueCheck.java:
```diff
@@ -0,0 +1,244 @@
+    super.visitFunctionCall(tree);
+  }
+
+  private static boolean parentUseValue(Tree child) {
```
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make "parentUseValue" a "static" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

pynicolas commented on php-frontend/src/main/java/org/sonar/php/tree/impl/statement/ForEachStatementTreeImpl.java:
```diff
@@ -157,6 +157,9 @@ public void accept(VisitorCheck visitor) {
   public static class ForEachStatementHeader implements Tree {
 
+    @Nullable
+    private Tree parent;
```
ForEachStatementHeader should not implement Tree.

pynicolas commented on php-frontend/src/main/java/org/sonar/plugins/php/api/tree/Tree.java:
```diff
@@ -666,6 +669,8 @@
      */
     HEREDOC_LITERAL(HeredocStringLiteralTree.class),
 
+    HEREDOC_BODY(HeredocBodyTree.class),
```
We don't need that.

pynicolas commented on php-frontend/src/main/java/org/sonar/plugins/php/api/tree/expression/HeredocBodyTree.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+import org.sonar.plugins.php.api.tree.Tree;
+
+public interface HeredocBodyTree extends Tree {
```
We don't need this interface.

pynicolas commented on php-frontend/src/main/java/org/sonar/php/tree/impl/expression/HeredocBodyTreeImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Override
+  public Kind getKind() {
+    return Kind.HEREDOC_BODY;
```
This method, as well as childrenIterator, can throw an exception.

pynicolas commented on php-checks/src/main/java/org/sonar/php/checks/UseOfEmptyReturnValueCheck.java:
```diff
@@ -0,0 +1,244 @@
+  private static boolean parentUseValue(Tree child) {
+    Tree parent = child.getParent();
+    if (parent == null) {
+      return false;
```
That's going to be hard to cover with unit tests.

alban-auzeill commented:
To keep the two commits clean, I didn't add a "Fix from review" commit, but I directly changed the existing commits:

- Add Tree#getParent()...
  Removed HeredocBodyTree interface
  Removed Tree.Kind.HEREDOC_BODY(HeredocBodyTree.class)

- SONARPHP-710 Rule: The ...
  Compare names using lower case: name.toLowerCase(Locale.ROOT)
  getParent() can never be null: Preconditions.checkNotNull(parent);
  move "nameOf(Tree tree)" into the class CheckUtils.getFunctionName(...)

alban-auzeill commented on php-frontend/src/main/java/org/sonar/php/tree/impl/expression/HeredocBodyTreeImpl.java:
```diff
@@ -0,0 +1,58 @@
+import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.php.api.visitors.VisitorCheck;
+
+public class HeredocBodyTreeImpl extends PHPTree {
```
Renamed into "HeredocBody" and to hide it more, I have moved "HeredocBody" as an inner class of "HeredocStringLiteralTreeImpl"
And moved all test of "HeredocBodyTreeTest"  into "HeredocStringLiteralTreeTest", so I can delete "HeredocBodyTreeTest"
